node/pkg/watchers/sui/watcher.go

```diff
@@ -503,25 +440,204 @@ func (e *Watcher) Run(ctx context.Context) error {
 
 	select {
 	case <-ctx.Done():
-		_ = ws.Close(websocket.StatusNormalClosure, "")
 		return ctx.Err()
 	case err := <-errC:
-		_ = ws.Close(websocket.StatusInternalError, err.Error())
 		return err
 	}
 }
 
-// fixSuiWsURL ensures the websocket scheme is properly specified
-func (e *Watcher) fixSuiWsURL(logger *zap.Logger) error {
-	u, _ := url.Parse(e.suiWS)
+func (w *Watcher) getEvents() ([]SuiResultInfo, error) {
+	// Only get events newer than the last processed height
+	var retVal []SuiResultInfo
+	var results []SuiResult
+	var txs []string
+	var nextCursor struct {
+		TxDigest string
+		EventSeq string
+	}
+	firstTime := true
+	for {
+		var payload string
+		if firstTime {
+			payload = w.queryEventsCmd
+			firstTime = false
+		} else {
+			payload = fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "suix_queryEvents", "params": [{ "MoveEventType": "%s" }, { "txDigest": "%s", "eventSeq": "%s" }, %d, %t]}`,
+				w.suiMoveEventType, nextCursor.TxDigest, nextCursor.EventSeq, w.maximumBatchSize, w.descendingOrder)
+		}
+		res, err := w.suiQueryEvents(payload)
+		if err != nil {
+			return retVal, err
+		}
+		for _, datum := range res.Result.Data {
+			txs = append(txs, *datum.ID.TxDigest)
+			results = append(results, datum)
+		}
+		if (len(res.Result.Data) == 0) || (len(txs) == 0) {
+			// In devnet (tilt) the core contract may not have any events and we don't want to flood the logs.
+			if w.unsafeDevMode {
+				return retVal, nil
+			}
+			return retVal, errors.New("getEvents was unable to get any events")
+		}
+		// Get and check the checkpoint for the last event against the lastProcessedHeight to see if we are done.
+		height, hErr := w.getCheckpointForDigest(txs[len(txs)-1])
+		if hErr != nil {
+			return retVal, hErr
+		}
+		if height <= w.latestProcessedCheckpoint || !res.Result.HasNextPage {
+			break
+		}
+		nextCursor.TxDigest = res.Result.NextCursor.TxDigest
+		nextCursor.EventSeq = res.Result.NextCursor.EventSeq
+	}
+	// At this point we have events but no checkpoints.
+	// Also, we probably have more events than we need.
+	// Need to do a bulk query to get all the checkpoints and then filter out the ones we don't need.
+	mbRes, err := w.getMultipleBlocks(txs)
+	if err != nil {
+		return retVal, err
+	}
+	if (len(mbRes) == 0) || (len(mbRes) != len(txs)) {
+		return retVal, errors.New("getEvents error getting multiple blocks")
+	}
+	for idx, block := range mbRes {
+		cp, err := strconv.ParseInt(block.Checkpoint, 10, 64)
+		if err != nil {
+			return retVal, fmt.Errorf("getEvents failed to ParseInt: %w", err)
+		}
+		if cp > w.latestProcessedCheckpoint {
+			// Can we double check the digests here?
```

Comment:
I feel like we can, since it's doing it here

Reply:
Removed